lxd/container_console.go

```diff
@@ -0,0 +1,483 @@
+			if err != nil {
+				return err
+			}
+
+			s.connsLock.Lock()
+			s.conns[fd] = conn
+			s.connsLock.Unlock()
+
+			if fd == -1 {
+				s.controlConnected <- true
+				return nil
+			}
+
+			s.connsLock.Lock()
+			for i, c := range s.conns {
+				if i != -1 && c == nil {
+					s.connsLock.Unlock()
+					return nil
+				}
+			}
+			s.connsLock.Unlock()
+
+			s.allConnected <- true
+			return nil
+		}
+	}
+
+	/* If we didn't find the right secret, the user provided a bad one,
+	 * which 403, not 404, since this operation actually exists */
+	return os.ErrPermission
+}
+
+func (s *consoleWs) Do(op *operation) error {
+	<-s.allConnected
+
+	var err error
+	master := &os.File{}
+	slave := &os.File{}
+	master, slave, err = shared.OpenPty(s.rootUid, s.rootGid)
+	if err != nil {
+		return err
+	}
+
+	if s.width > 0 && s.height > 0 {
+		shared.SetSize(int(master.Fd()), s.width, s.height)
+	}
+
+	controlExit := make(chan bool)
+	var wgEOF sync.WaitGroup
+
+	consolePidChan := make(chan int)
+	wgEOF.Add(1)
+	go func() {
+		select {
+		case <-s.controlConnected:
+			break
+
+		case <-controlExit:
+			return
+		}
+
+		consolePid := <-consolePidChan
+
+		for {
+			s.connsLock.Lock()
+			conn := s.conns[-1]
+			s.connsLock.Unlock()
+
+			_, r, err := conn.NextReader()
+			if err != nil {
+				logger.Debugf("Got error getting next reader %s", err)
+				err := syscall.Kill(consolePid, syscall.SIGTERM)
+				if err != nil {
+					logger.Debugf("Failed to send SIGTERM to pid %d", consolePid)
+				} else {
+					logger.Debugf("Sent SIGTERM to pid %d", consolePid)
+				}
+				return
+			}
+
+			buf, err := ioutil.ReadAll(r)
+			if err != nil {
+				logger.Debugf("Failed to read message %s", err)
+				break
+			}
+
+			command := api.ContainerConsoleControl{}
+
+			err = json.Unmarshal(buf, &command)
+			if err != nil {
+				logger.Debugf("Failed to unmarshal control socket command: %s", err)
+				continue
+			}
+
+			if command.Command == "window-resize" {
+				winchWidth, err := strconv.Atoi(command.Args["width"])
+				if err != nil {
+					logger.Debugf("Unable to extract window width: %s", err)
+					continue
+				}
+
+				winchHeight, err := strconv.Atoi(command.Args["height"])
+				if err != nil {
+					logger.Debugf("Unable to extract window height: %s", err)
+					continue
+				}
+
+				err = shared.SetSize(int(master.Fd()), winchWidth, winchHeight)
+				if err != nil {
+					logger.Debugf("Failed to set window size to: %dx%d", winchWidth, winchHeight)
+					continue
+				}
+
+				logger.Debugf("Set window size to: %dx%d", winchWidth, winchHeight)
+			}
+		}
+	}()
+
+	go func() {
+		s.connsLock.Lock()
+		conn := s.conns[0]
+		s.connsLock.Unlock()
+
+		logger.Debugf("Starting to mirror websocket")
+		readDone, writeDone := shared.WebsocketConsoleMirror(conn, master, master)
+
+		<-readDone
+		logger.Debugf("Finished to read websocket")
+		<-writeDone
+		logger.Debugf("Finished to write websocket")
+		logger.Debugf("Finished to mirror websocket")
```

Comment:
That looks like possibly excessive logging here? especially for those of us always running LXD in debug mode :)

Reply:
No, that only logs on disconnect which happens once.

Reply:
ok